lib/defender/aws_iot_defender.c

```diff
@@ -0,0 +1,183 @@
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <stdbool.h>
+
+/* Defender internal include. */
+#include "private/aws_iot_defender_internal.h"
+
+#include "platform/aws_iot_network.h"
+
+#include "platform/aws_iot_clock.h"
+
+#define _START_INFO_RESETER    ( AwsIotDefenderStartInfo_t ) { 0 }
+
+/* Period between reports in milliseconds. */
+static uint32_t _periodMilliSecond = AWS_IOT_DEFENDER_DEFAULT_PERIOD_SECONDS * 1000;
+
+/* the timer that connect, subscribe, publish metrics report */
+static AwsIotTimer_t _metricsPublishTimer;
+
+/* the timer that disconnect */
+static AwsIotTimer_t _guardTimer;
+
+static AwsIotMqttConnection_t _mqttConnection = AWS_IOT_MQTT_CONNECTION_INITIALIZER;
+
+static AwsIotDefenderStartInfo_t _startInfo = { 0 };
+
+/**
+ * Return the state that whether defender is already started.
+ */
+static bool isStarted();
+
+/**
+ * Create and arm _metricsPublishTimer
+ */
+static void _metricsPublishTimerExpirationRoutine( void * );
+static void _guardTimerExpirationRoutine( void * );
+
+/*-----------------------------------------------------------*/
+
+static bool isStarted()
+{
+    return _startInfo.pThingName != NULL && _startInfo.thingNameLength > 0;
+}
+
+/*-----------------------------------------------------------*/
+
+AwsIotDefenderError_t AwsIotDefender_Start( AwsIotDefenderStartInfo_t * pStartInfo )
+{
+    AwsIotDefenderError_t defenderError = AWS_IOT_DEFENDER_INTERNAL_FAILURE;
+
+    /* if current state is not started */
+    if( !isStarted() )
+    {
+        /* copy input start info into global variable _startInfo */
+        _startInfo = *pStartInfo;
+
+        /* metrics publish timer must be created and armed */
+        /* guard timer must be created */
+        if( AwsIotClock_TimerCreate( &_metricsPublishTimer, _metricsPublishTimerExpirationRoutine, NULL ) )
+        {
+            if( AwsIotClock_TimerArm( &_metricsPublishTimer, _periodMilliSecond, _periodMilliSecond ) )
+            {
+                if( AwsIotClock_TimerCreate( &_guardTimer, _guardTimerExpirationRoutine, NULL ) )
+                {
+                    defenderError = AWS_IOT_DEFENDER_SUCCESS;
+                }
```

Comment:
missing else?

Reply:
No. The return value is initialized AWS_IOT_DEFENDER_INTERNAL_FAILURE. So by default, this function return "internal failure"